ddprof-lib/src/main/cpp/wallClock.h

```diff
@@ -19,59 +19,152 @@
 
 #include "engine.h"
 #include "os.h"
+#include "profiler.h"
+#include "reservoirSampler.h"
+#include "threadFilter.h"
 #include "threadState.h"
-#include <atomic>
-#include <climits>
-#include <pthread.h>
-#include <signal.h>
-
-class WallClock : public Engine {
-private:
-  static std::atomic<bool> _enabled;
-
-  bool _collapsing;
-  long _interval;
-
-  // Maximum number of threads sampled in one iteration. This limit serves as a
+#include "tsc.h"
+#include "vmStructs.h"
+
+class BaseWallClock : public Engine {
+  private:
+    static std::atomic<bool> _enabled;
+    std::atomic<bool> _running;
+  protected:
+    long _interval;
+    // Maximum number of threads sampled in one iteration. This limit serves as a
   // throttle when generating profiling signals. Otherwise applications with too
   // many threads may suffer from a big profiling overhead. Also, keeping this
   // limit low enough helps to avoid contention on a spin lock inside
   // Profiler::recordSample().
   int _reservoir_size;
 
-  std::atomic<bool> _running;
-  pthread_t _thread;
-
-  void timerLoop();
+    pthread_t _thread;
+    virtual void timerLoop() = 0;
+    virtual void initialize(Arguments& args) {};
 
   static void *threadEntry(void *wall_clock) {
-    ((WallClock *)wall_clock)->timerLoop();
+    ((BaseWallClock *)wall_clock)->timerLoop();
     return NULL;
   }
 
-  static bool inSyscall(void *ucontext);
-
-  static void sharedSignalHandler(int signo, siginfo_t *siginfo,
-                                  void *ucontext);
-  void signalHandler(int signo, siginfo_t *siginfo, void *ucontext,
-                     u64 last_sample);
+    bool isEnabled() const;
+
+    template <typename ThreadType, typename CollectThreadsFunc, typename SampleThreadsFunc>
+    void timerLoopCommon(CollectThreadsFunc collectThreads, SampleThreadsFunc sampleThreads, int reservoirSize, u64 interval) {
+        if (!_enabled.load(std::memory_order_acquire)) {
+            return;
+        }
+
+        // Dither the sampling interval to introduce some randomness and prevent step-locking
+        const double stddev = ((double)_interval) / 10.0;  // 10% standard deviation
+        // Set up random engine and normal distribution
+        std::random_device rd;
```

Comment:
Shouldn't this stuff use the ReservoirSampler class?

Reply:
Yes, it does. The random thingy here is to dither the sampling period.

Reply:
Yes, I realised after commenting